### PR TITLE
feat: modemManager based position provider

### PR DESCRIPTION
**Pull request by sfiorani**
This PR introduces a new PositionProvider based on the ModemManager location. Location is retrieved from the Modem with Dbus APIs, from the RAW and NMEA location source of the modem. Obviously, this provider requires to have modem manager installed and running on the device.

The provider is located under the NM bundle, as we've direct access to the dbus api, and only some exports from the linux.position bundle are required.

**Related Issue:** This PR fixes/closes {issue number}

**Description of the solution adopted:** A more detailed description of the changes made to solve/close one or more issues. If the PR is simple and easy to understand this section can be skipped

**Screenshots:** If applicable, add screenshots to help explain your solution

**Manual Tests**: Optional description of the tests performed to check correct functioning of changes, useful for an efficient review

**Any side note on the changes made:** Description of any other change that has been made, which is not directly linked to the issue resolution [e.g. Code clean up/Sonar issue resolution]

### CODE REVIEW

**Comment by mattdibi** [comment]
@sfiorani please add a description for this PR

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/ModemManagerDbusWrapper.java [hunk 1]
```diff
@@ -40,17 +40,19 @@ public class ModemManagerDbusWrapper {
 
     private static final Logger logger = LoggerFactory.getLogger(ModemManagerDbusWrapper.class);
 
-    private static final String MM_BUS_NAME = "org.freedesktop.ModemManager1";
+    private static final String NETWORK_MANAGER_BUS_NAME = "org.freedesktop.NetworkManager";
+    private static final String MM_DEVICE_NAME = "org.freedesktop.NetworkManager.Device";
```
Comment (pierantoniomerlino) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/ModemManagerDbusWrapper.java]: Both `NETWORK_MANAGER_BUS_NAME` and `MM_DEVICE_NAME` are not used.

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/ModemManagerDbusWrapper.java [hunk 2]
```diff
@@ -40,17 +40,19 @@ public class ModemManagerDbusWrapper {
...
+    private static final String MODEM_MANAGER_BUS_NAME = "org.freedesktop.ModemManager1";
```
Comment (pierantoniomerlino) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/ModemManagerDbusWrapper.java]: Following the convention already used in this class, the name should be `MM_BUS_NAME`

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/ModemManagerDbusWrapper.java [hunk 3]
```diff
@@ -126,9 +128,27 @@ protected void enableModem(String modemDevicePath) throws DBusException {
...
     }
 
-    protected Optional<Properties> getModemProperties(String modemPath) throws DBusException {
+    public Modem getModem(String modemPath) throws DBusException {
+        return dbusConnection.getRemoteObject(MODEM_MANAGER_BUS_NAME, modemPath, Modem.class);
+    }
+
+    public Location getModemManagerLocation(String modemPath) throws DBusException {
+        return dbusConnection.getRemoteObject("org.freedesktop.ModemManager1", modemPath, Location.class);
+    }
```
Comment (pierantoniomerlino) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/ModemManagerDbusWrapper.java:135]: Please use the constant already defined for the Modem path.

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/ModemManagerDbusWrapper.java [hunk 4]
```diff
@@ -126,9 +128,27 @@ protected void enableModem(String modemDevicePath) throws DBusException {
...
+
+    public Properties getLocationProperties(Location location) throws DBusException {
+        return dbusConnection.getRemoteObject("org.freedesktop.ModemManager1", location.getObjectPath(),
+                Properties.class);
```
Comment (pierantoniomerlino) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/ModemManagerDbusWrapper.java]: Also here.

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/NMDbusConnector.java [hunk 5]
```diff
@@ -642,4 +646,55 @@ private void configurationEnforcementDisable() throws DBusException {
         logger.debug("Network configuration enforcement set to {} (Expected: false)",
                 this.configurationEnforcementHandlerIsArmed);
     }
+
+    public List<Location> getAvailableMMLocations() {
+        List<Location> availableLocations = new ArrayList<>();
+
+        this.getEnabledModemsPaths().stream().forEach(modemPath -> {
```
Comment (pierantoniomerlino) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/NMDbusConnector.java]: I think that you can use the `forEach` method directly to the List, without using `stream()`

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/position/MMLocationParser.java [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+    private Double speed;
+    private Double track;
+
+    private int fixQuality;
+    private int nrSatellites;
+    private double mDOP;
+    private double mPDOP;
+    private double mHDOP;
+    private double mVDOP;
+    private int m3Dfix;
```
Comment (pierantoniomerlino) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/position/MMLocationParser.java:57]: This field is not used or assigned.

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/position/MMLocationParser.java [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+    private char latitudeHemisphere;
+    private char longitudeHemisphere;
+    private char validFix = 0;
```
Comment (pierantoniomerlino) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/position/MMLocationParser.java:60]: These three fields are assigned, but not used.

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/position/MMLocationParser.java [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+                    parseGgaSentence(tokens);
+                    break;
+
+                default:
+                    // Do Nothing
+
+                }
+            }
+
+        } catch (Exception ex) {
```
Comment (pierantoniomerlino) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/position/MMLocationParser.java:166]: Here you catch all the exceptions, but the code above does not throw anything. Do you implemented this for catching the unchecked exceptions, like the `NumberFormatException`?

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/position/MMPositionProvider.java [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+import org.freedesktop.dbus.types.UInt32;
+import org.freedesktop.dbus.types.Variant;
+import org.freedesktop.modemmanager1.modem.Location;
+import org.osgi.util.position.Position;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MMPositionProvider implements PositionProvider {
+
+    private static final Logger logger = LoggerFactory.getLogger(MMPositionProvider.class);
```
Comment (pierantoniomerlino) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/position/MMPositionProvider.java:43]: This logger is not used

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/position/MMPositionProvider.java [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+    private static final UInt32 NMEA_LOCATION_SOURCE = new UInt32(
+            MMModemLocationSource.MM_MODEM_LOCATION_SOURCE_GPS_NMEA.getValue());
+    private static final UInt32 RAW_LOCATION_SOURCE = new UInt32(
+            MMModemLocationSource.MM_MODEM_LOCATION_SOURCE_GPS_RAW.getValue());
+
+    private DateTimeFormatter nmeaDateTimePattern = DateTimeFormatter.ofPattern("ddMMyy hhmmss");
```
Comment (pierantoniomerlino) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/position/MMPositionProvider.java]: Also this one is not used.

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/position/MMPositionProvider.java [hunk 11]
```diff
@@ -0,0 +1,158 @@
+/*******************************************************************************
```
Comment (pierantoniomerlino) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/position/MMPositionProvider.java:1]: Please run the clean up to this class.

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/position/MMPositionProvider.java [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+    public MMPositionProvider(NMDbusConnector connector) {
+        this.nmDbusConnector = connector;
+    }
+
+    @Override
+    public void start() {
+
+        this.positionRefreshTask = Executors.newSingleThreadScheduledExecutor();
+        this.positionRefreshTask.scheduleAtFixedRate(this::getModemManagerLocation, 0, refreshRate, TimeUnit.SECONDS);
+    }
```
Comment (pierantoniomerlino) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/position/MMPositionProvider.java:78]: It is supposed that the `init` method is called before the `start` one. But what happens if this is not done? The `refresh` field is initializated to 0 and the schedule throws an `IllegalArgumentException`. Maybe it is better to add a check of this variable before using it.

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/position/MMPositionProvider.java [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+        throw new UnsupportedOperationException("LastSentence not available on ModemManagaer provider");
+    }
+
+    @Override
+    public void init(PositionServiceOptions configuration, LockStatusListener gpsDeviceListener,
+            GpsDeviceAvailabilityListener gpsDeviceAvailabilityListener) {
+
+        this.gpsDeviceListener = gpsDeviceListener;
+        this.refreshRate = configuration.getModemManagerRefreshRate();
+
```
Comment (pierantoniomerlino) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/position/MMPositionProvider.java:133]: The `gpsDeviceAvailabilityListener` is not used here. Is it correct?

**Review by pierantoniomerlino — CHANGES_REQUESTED** [review]

**Comment by MMaiero** [comment]
image

@sfiorani

**Review by pierantoniomerlino — APPROVED** [review]

**Comment by pierantoniomerlino** [comment]
@MMaiero the Sonar analysis failed, but @sfiorani only renamed the files. I think that we shouldn't take care of this refactor in this PR.

**Pull request merged**